src/BuildingRegistry.Importer.Console/CommandGenerator.cs

```diff
@@ -258,12 +258,33 @@ public IEnumerable<dynamic> GenerateUpdateCommandsFor(int key, DateTime @from, D
 
             if (importMode == ImportMode.Update)
             {
-                var houseNumbersForUpdate = importTerrainObjectHouseNumberCommands
-                    .Where(x => x.Timestamp > from.ToCrabInstant() && x.Timestamp <= until.ToCrabInstant())
-                    .Select(x => x.HouseNumberId).ToList();
+                var importTerrainObjectHouseNumberFromCrabInRange = importTerrainObjectHouseNumberCommands
+                    .Where(x => x.Timestamp > @from.ToCrabInstant() && x.Timestamp <= until.ToCrabInstant())
+                    .ToList();
+
+                var houseNumbersForUpdate = importTerrainObjectHouseNumberFromCrabInRange.Select(x => x.HouseNumberId).ToList();
 
                 if (houseNumbersForUpdate.Any())
                 {
+                    var terrainObjectHouseNumberHouseNrCombinationsBeforeToday = importTerrainObjectHouseNumberCommands
+                        .Where(x => x.Timestamp <= from.ToCrabInstant())
+                        .Select(x => Tuple.Create(x.TerrainObjectHouseNumberId, x.HouseNumberId))
+                        .Distinct()
+                        .ToList();
+
+                    foreach (var terrainObjectHouseNumberFromCrab in importTerrainObjectHouseNumberFromCrabInRange.Where(x => !terrainObjectHouseNumberHouseNrCombinationsBeforeToday.Any(tuple => tuple.Item1 == x.TerrainObjectHouseNumberId && tuple.Item2 == x.HouseNumberId)))
```

Comment:
extract that where clause from the foreach difinition. It's not a simple filter and makes it very hard to read what you are filtering.

Reply:
use meaningful names in the lambda
`x, tuple, item1, item2` make it very hard to read what's being compared

Reply:
Can't extract where clause when using anonymous objects. I'll reformat to make it more readable